Add/update alt text in documentation (#607)
jupyter/jupyter: docs/source/projects/architecture/content-architecture.rst

> The core execution machinery for the kernel is shared with terminal IPython:

Reviewer: Question about a probably very minor issue: is the fact that there is a colon at the end of the sentence somehow conveyed by the screen reader? In which case, would the document flow for the screen reader user be somewhat disrupted if the phrase preceding the colon is not semantically connected to the phrase following the colon?

In other words, I'm concerned that to a screen reader, we have

> The core execution machinery for the kernel is shared with terminal IPython:
> A kernel process can be connected to more than one frontend simultaneously. In
> this case, the different frontends will have access to the same variables.

But it doesn't make sense to join those two statements with a colon. It makes sense to join-via-colon the first sentence to the image that it describes/captions.

This colon-terminated phrase is really a caption for the image. But the whole figure-caption machinery as of now creates a bad user experience. So I wonder what a good solution would be.

Author: Is there a reason not to change it to a period? Maybe I'm missing something.

> The core execution machinery for the kernel is shared with terminal IPython:

Reviewer: ```suggestion
The core execution machinery for the kernel is shared with terminal IPython.
```

The core execution machinery for the kernel is shared with terminal IPython:

.. image:: figs/ipy_kernel_and_terminal.png

>    :alt: 

Author: Good thought! I wasn't sure if I was thinking too far ahead, but if we both are then I think it's okay. 😆

>    :alt: 

Author: Did you want them added only to the top of each document? As of now I have comments below each empty alt.

>    :alt: Venn diagram showing that Terminal IPython and the IPython Kernel both share the same Python executable. Terminal IPython communicates with the Python executable using stdin and stdout, whereas the Python Kernel uses JSON messages sent over ZeroMQ. 

Reviewer 2: Is there a best practice for including information as alt text when that information would *also* be useful as "regular" text that comes just before or after the image? Some of these alt-texts actually read like captions which would be useful regardless of whether the reader is using assistive technology. If we defined these as captions instead of `alt-text` (or used very simple alt-text in addition to captions) would that be an option from an a11y perspective?

Author: This is a great question! I've been having this discussion with a few different people recently, but I'm not aware of any best practices written for this yet since what is or isn't  included *outside* of the image is often considered the realm of the documentation's writing style guide (or lack thereof). 

There are, however, resources on how to combine captions and alt text that we could use if we go that direction. Basically, if we'd rather have this as body text or a caption and not alt text, we want to have an empty `alt` element for the images (which means they are skipped). Here's the [W3C recommendation on images with adjacent text](https://www.w3.org/WAI/tutorials/images/decorative/#image-with-adjacent-text-alternative).

There's also been some discussion around whether we want to use [`aria-describedby`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) to indicate that there is an image while explicitly linking to its explanation. This could be helpful if the related information is not directly next to the image. I personally think this could be a misuse of the tag, though, especially when our goal is to use proper HTML over adding ARIA. 

cc @tonyfast, @gabalafou, and @ohrely (in case you all are interested)

---

Overall, I'm open to making some of these alt text contributions captions of some kind in this or a future PR. My main goal is to give these sprint contributors credit (via commit) however we decide to do so.

Reviewer 2: Hmmm - here's an idea: what if we converted these into figures with captions, instead of images. [Here's an example of what they would look like from Jupyter Book](https://jupyterbook.org/content/figures.html?highlight=figures#figures).

There's a directive for this in Sphinx. Basically, this image directive:

```rst
.. image:: figs/ipy_kernel_and_terminal.png
   :alt: Venn diagram showing that Terminal IPython and the IPython Kernel both share the same Python executable. Terminal IPython communicates with the Python executable using stdin and stdout, whereas the Python Kernel uses JSON messages sent over ZeroMQ. 
```

would become this figure directive:

```rst
.. figure:: figs/ipy_kernel_and_terminal.png
   
   Venn diagram showing that Terminal IPython and the IPython Kernel both share the same Python executable. Terminal IPython communicates with the Python executable using stdin and stdout, whereas the Python Kernel uses JSON messages sent over ZeroMQ. 
```

Reviewer 2: Here's an example from the Jupyter Book docs:

https://jupyterbook.org/en/stable/content/figures.html#figures

Here's a sample HTML snippet:

```html
<div class="figure align-default" id="directive-fig">
<a class="reference internal image-reference" href="../_images/C-3PO_droid.png"><img alt="../_images/C-3PO_droid.png" src="../_images/C-3PO_droid.png" style="height: 150px;"></a>
<p class="caption"><span class="caption-number">Fig. 4 </span><span class="caption-text">Here is my figure caption!</span><a class="headerlink" href="#directive-fig" title="Permalink to this image">¶</a></p>
</div>
```

If that's not a11y-friendly I'm happy to just go with alt tags

Reviewer: My recommendation is that we remove the alt text of the diagrams and review them. By remove, I mean set the alt text to empty. By review, I mean we should make sure that the surrounding text describes the diagram. In other words, we should make sure the diagrams do not contain any information that isn't already conveyed by reading the article without the diagrams. I want to be clear that I am only referring to diagram-type images, not all images.

Here's why. 

I did a deep dive into this subject this afternoon. I dug through threads on the global accessibility Slack. I read web pages on the W3C and Gov.UK sites. I also tested with my own hardware using combinations of VoiceOver and NVDA with Safari, Chrome and Edge. 

The main problem is that there are cross-screen-reader compatibility issues. Not all screen readers treat the figure tag the same way. 

The other problem is that my alt-text diagram descriptions typically repeat or paraphrase stuff that's already in the article. So unless we make my descriptions redundant/repetitive for all users, I don't think we should potentially degrade the screen reader experience by making it more repetitive and verbose.

This seems to align with the [UK government's guidance for describing images.](https://www.gov.uk/guidance/content-design/images)

If we agree with this course of action, then I will review the PR and use the suggestion feature to make the changes I that I am recommending.

Reviewer 2: I defer to your and others expertise when it comes to the vest practices for accessibility. Though it sounds like some of the changes you suggest would also just be more generally useful for readers, if we are removing duplicate information and such. So I'm extra +1 on that as well.

> in the target language:

Reviewer: ```suggestion
in the target language.
```

``.ipynb`` extension.

.. image:: figs/notebook_components.png

>    :alt: Communication diagram showing five nodes labeled user, browser, notebook server, notebook file, and kernel. The notebook server is a communication hub. The browser, notebook file and kernel cannot talk to each other directly. They communicate indirectly through the notebook server. The user can only talk to the browser. The notebook server and kernel are colored green whereas the browser is colored purple to draw a distinction between the Jupyter front end versus back end.

Reviewer: ```suggestion
   :alt:
   
.. 
   Alt text is intentionally left blank because the image content is described thoroughly in the surrounding text.

The notebook server is a communication hub. The browser, notebook file on disk, and
kernel cannot talk to each other directly. They communicate through the notebook server. 
```

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

steps:

.. image:: figs/nbconvert.png

>    :alt: 

Reviewer: agree

.. image:: /_static/_images/repos_map.png
   :width: 75%

>    :alt: Architecture diagram of Jupyter project relationships from servers, applications, API, and kernels.

Reviewer: Good lord. How do we make such a dense relationship map accessible? I feel like this is getting into new territory for me

Author: I personally think the best option would be to have this be a part of the main text with an empty alt attribute, but I know I do not have the time to write it at the moment. My current guess is that it would make sense to start by describing the overlapping content first to minimize repetition, then call out areas by section. 

Maybe it's the wrong attitude, but I was going with "something is better than nothing" for this. If it's a blocker, let me know and I'll find a way to get it done.

If we think the image is the way to go, I think this one needs a linked long description. Even that will be tricky with all the relationships!

Reviewer: Agreed. This will need to be revisited in the future.
